src/metabase/util/malli.clj

```diff
@@ -87,16 +88,24 @@ explain-fn-fail!
   [:and any?
    [:fn {:description "a malli schema"} mc/schema]])
 
-(defn with-api-error-message :- Schema
+(def ^:private localized-string-schema
+  [:fn {:error/message "must be a localized string"}
+   i18n/localized-string?])
+
+(defn with-api-error-message
   "Update a malli schema to have a :description (picked up by api docs),
-  and a :error/message (used by defendpoint). They don't have to be the same, but usually are."
-  ([mschema :- Schema message :- string?]
-   (with-api-error-message mschema message message))
-  ([mschema :- Schema
-    docs-message :- string?
-    error-message :- string?]
-   (mut/update-properties mschema assoc
-                          ;; override generic description in api docs
-                          :description docs-message
-                          ;; override generic description in defendpoint api errors
-                          :error/message error-message)))
+  and a :error/message (used by defendpoint). They don't have to be the same, but usually are.
+
+    (with-api-error-message
+      [:string {:min 1}]
+      (deferred-tru \"Must be a string with at least 1 character.\"))"
+  ([mschema :- Schema error-message :- localized-string-schema]
+   (with-api-error-message mschema error-message error-message))
+  ([mschema         :- :any
+    error-message   :- localized-string-schema
+    specific-errors :- localized-string-schema]
+   (mut/update-properties (mc/schema mschema) assoc
+                          ;; override generic description in api docs and :errors in API's response
+                          :description error-message
+                          ;; override generic description in :specific-errors in API's response
+                          :error/fn    (fn [_ _] specific-errors))))
```

Comment:
Overall I think it's a great change.

I can't leave a "suggestion" because it spans a deleted line (or something), but here's what I would suggest:

```clojure
(defn with-api-error-message
  "Update a malli schema to have a :description (used by umd/describe, which is used by api docs),
  and a :error/fn (used by me/humanize, which is used by defendpoint).
  They don't have to be the same, but usually are.

  (with-api-error-message
    [:string {:min 1}]
    (deferred-tru \"Must be a string with at least 1 character representing a User ID.\"))"
  ([mschema :- Schema error-message :- localized-string-schema]
   (with-api-error-message mschema error-message error-message))
  ([mschema                :- :any
    description-message    :- localized-string-schema
    specific-error-message :- localized-string-schema]
   (mut/update-properties (mc/schema mschema) assoc
                          ;; override generic description in api docs and :errors key in API's response
                          :description description-message
                          ;; override generic description in :specific-errors key in API's response
                          :error/fn    (fn [_ _] specific-error-message))))

```

some tests worth running:

```clojure
(comment
  (require '[malli.core :as mc] ;; nocommit
           '[malli.error :as me]
           '[malli.util :as mut]
           '[metabase.util.malli :as mu]
           '[metabase.util.malli.describe :as umd] ;; umd/describe
           '[malli.provider :as mp]
           '[malli.generator :as mg]
           '[malli.transform :as mtx])

  (require '[metabase.util.i18n :refer [deferred-tru trs tru]])

  (let [description (deferred-tru "this is a description that is used to explain what shape is required. The shape of an input is not used to create me.")
        specific-error (deferred-tru "this is an error explaination that will be tied to an invalid input.")
        add-my-errors (fn [schema] (with-api-error-message schema description specific-error))]

    [(= (str description) (umd/describe (add-my-errors :int)))
     (= [specific-error] (me/humanize (mc/explain (add-my-errors :int) "3")))
     (= [specific-error] (me/humanize (mc/explain (add-my-errors [:map [:x :int]]) "3")))
     (= ["this is an error explaination that will be tied to an invalid input. <-> 3"]
        (me/humanize (mc/explain (add-my-errors [:map [:x :int]]) "3")
                     {:wrap #(str (:message %) " <-> " (:value %))}))]
    )

  )
  ```
  
I think it's also worth adding a test that uses with-api-error-message into `metabase.api.common.internal-test`, since that's a bit closer to what will actually happen on an api call.

Reply:
agreed on adding a test for api message, I added it here: 1679bb654b

Tho the test you recommend above looks nice, but I don't see how that differs from what we're having in `test/metabase/util/malli_test.clj`?